allennlp/data/tokenizers/pretrained_transformer_tokenizer.py

```diff
@@ -22,36 +24,98 @@ class PretrainedTransformerTokenizer(Tokenizer):
+    stride: ``int``, optional (default=0)
+        If set to a number along with max_length, the overflowing tokens returned will contain some tokens
+        from the main sequence returned. The value of this argument defines the number of additional tokens.
+    truncation_strategy: ``str``, optional (default='longest_first')
+        String selected in the following options:
+        - 'longest_first' (default) Iteratively reduce the inputs sequence until the input is under max_length
+        starting from the longest one at each token (when there is a pair of input sequences)
+        - 'only_first': Only truncate the first sequence
+        - 'only_second': Only truncate the second sequence
+        - 'do_not_truncate': Does not truncate (raise an error if the input sequence is longer than max_length)
+
+    Argument descriptions are from
+    https://github.com/huggingface/transformers/blob/155c782a2ccd103cf63ad48a2becd7c76a7d2115/transformers/tokenization_utils.py#L691
     """
 
     def __init__(
-        self, model_name: str, start_tokens: List[str] = None, end_tokens: List[str] = None
+        self,
+        model_name: str,
+        start_tokens: List[str] = None,
+        end_tokens: List[str] = None,
+        add_special_tokens: bool = True,
+        max_length: int = None,
+        stride: int = 0,
+        truncetion_strategy: str = "longest_first",
     ) -> None:
         self._tokenizer = AutoTokenizer.from_pretrained(model_name)
-        default_start_tokens, default_end_tokens = _guess_start_and_end_token_defaults(model_name)
-        self._start_tokens = start_tokens if start_tokens is not None else default_start_tokens
-        self._end_tokens = end_tokens if end_tokens is not None else default_end_tokens
+        self._start_tokens = start_tokens
+        self._end_tokens = end_tokens
+        self._add_special_tokens = add_special_tokens
+        self._max_length = max_length
+        self._stride = stride
+        self._truncation_strategy = truncetion_strategy
+
+    def _add_custom_start_end_tokens(self, text: str) -> str:
+        if self._start_tokens is not None:
+            text = " ".join(self._start_tokens) + " " + text
+        if self._end_tokens is not None:
+            text = text + " " + " ".join(self._end_tokens)
+        return text
 
     @overrides
     def tokenize(self, text: str) -> List[Token]:
+        """
+        This can also handle sentence pairs. In this case 1st and 2nd sentences in ``text``
+        are expected to be separated with ``ALLENNLP_SENTENCE_PAIR_SEP`` symbol.
+        """
         # TODO(mattg): track character offsets.  Might be too challenging to do it here, given that
         # ``transformers``` is dealing with the whitespace...
-        token_strings = self._start_tokens + self._tokenizer.tokenize(text) + self._end_tokens
-        return [Token(t) for t in token_strings]
+        if ALLENNLP_SENTENCE_PAIR_SEP in text:
```

Comment:
I'd be happier if we added a `tokenize_sentences(sentences: List[str]) -> List[Token]` method to `Tokenizer`, instead of adding this special constant.  But I suppose this could do for now...  I'm just wary of adding a recommended public constant that we might want to remove later.

Reply:
Yeah, but the downside of it is that we cannot handle a sentence pair with an inherited ``tokenize`` method in this setup. So I introduced this symbol to tell this method that it gets a sentence pair. Optional argument to the ``tokenize`` might be a cleaner solution, but this breaks API. 

I am not sure however how important it is to handle a sentence pair in the main tokenize method and I followed your suggestion.

Regarding 3+ sentences, to the best of my knowledge, it will be a custom logic that is not available in transformers repo. Probably for another PR.

Reply:
Well, either way the `DatasetReader` has to do something differently to handle this.  In your version, you have to manually add the separator token, then have the tokenizer immediately undo this concatenation and process things accordingly.  This also means that we'd have to modify *every* tokenizer to handle this token, because the `DatasetReader` would be adding it, and our other tokenizers would see it if you switched.

In my way, the default behavior is just to concatenate the sentences (probably with a space) and pass it to `tokenize()`.  No other tokenizers break, and so the `DatasetReader` doesn't have to have custom logic that's aware of which `tokenizer` it's using.

Reply:
I see your point regarding the special token, it totally makes sense to omit it since we might want to change the tokenizer type we are using.

Regarding concatenating sentence pair with space, how will you then decide where to split it in ``tokenize``? Huggingface's ``encode_plus`` method expects sentences to come separated.

Reply:
I was talking about the default implementation in `Tokenizer.tokenize_sentences`, which should be easy.  `PretrainedTransformerTokenizer` would override `tokenize_sentences` with the logic you have here.